kernel/src/main/scala/cats/kernel/Enum.scala

```diff
@@ -0,0 +1,85 @@
+package cats
+package kernel
+
+import scala.annotation.tailrec
+import scala.{specialized => sp}
+
+/**
+ * A typeclass with an operation which returns a member which is
+ * greater or `None` than the one supplied.
+ */
+trait PartialNext[@sp A] {
+  def partialOrder: PartialOrder[A]
+  def partialNext(a: A): Option[A]
+}
+
+/**
+ * A typeclass with an operation which returns a member which is
+ * always greater than the one supplied.
+ */
+trait Next[@sp A] extends PartialNext[A] {
+  def next(a: A): A
+  override def partialNext(a: A): Option[A] = Some(next(a))
+}
+
+/**
+ * A typeclass with an operation which returns a member which is
+ * smaller or `None` than the one supplied.
+ */
+trait PartialPrevious[@sp A] {
+  def partialOrder: PartialOrder[A]
+  def partialPrevious(a: A): Option[A]
+}
+
+/**
+ * A typeclass with an operation which returns a member which is
+ * always smaller than the one supplied.
+ */
+trait Previous[@sp A] extends PartialPrevious[A] {
+  def partialOrder: PartialOrder[A]
+  def previous(a: A): A
+  override def partialPrevious(a: A): Option[A] = Some(previous(a))
+}
+
+/**
+ * A typeclass which has both `previous` and `next` operations
+ * such that `next . previous == identity`.
+ */
+trait UnboundedEnum[@sp A] extends Next[A] with Previous[A] {
+  def order: Order[A]
+  override def partialOrder: PartialOrder[A] = order
+}
+
+trait BoundedEnum[@sp A] extends PartialPrevious[A] with PartialNext[A] with UpperBounded[A] with LowerBounded[A] {
+
+  def order: Order[A]
+  override def partialOrder: PartialOrder[A] = order
+
+  def cycleNext(a: A): A =
+    partialNext(a).getOrElse(minBound)
+
+  def cyclePrevious(a: A): A =
+    partialPrevious(a).getOrElse(maxBound)
+
+  /** Enumerate the members in ascending order. */
+  def members: List[A] = {
```

Comment:
Should this maybe be an `Iterable` or `LazyList` / `Stream`? Having a generic method that returns all members as a list in a context where instances will often be for types like `Int` seems not really ideal.

Reply:
I agree! I toyed around with `Iterable`, but didn't like the idea because it's not pure iirc, and the idea of having an impure return type irked me. I don't have a strong preference on either lazy types but I think a `NonEmptyLazyList` sounds good? I can then add `membersAscending`/`membersDescending` on the `LowerBounded` and `UpperBounded` `Enum` types.